scheduler/scheduler.go

```diff
@@ -0,0 +1,111 @@
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+)
+
+type job struct {
+	cmd      func()
+	name     string
+	schedule schedule
+}
+
+type Scheduler struct {
+	cv       *sync.Cond
+	jobs     []job
+	logger   *log.Logger
+	mutex    *sync.Mutex
+	numReady int
+	ready    sync.Map
+}
+
+func New(logger *log.Logger) *Scheduler {
+	mutex := &sync.Mutex{}
+	cv := sync.NewCond(mutex)
+
+	return &Scheduler{
+		cv:       cv,
+		jobs:     nil,
+		logger:   logger,
+		mutex:    mutex,
+		numReady: 0,
+		ready:    sync.Map{},
+	}
+}
+
+func (s *Scheduler) Add(schedule string, cmd func()) (bool, error) {
+	return s.add(schedule, cmd, fmt.Sprintf("job-%d", time.Now().UTC().Unix()))
+}
+
+func (s *Scheduler) AddWithName(schedule string, cmd func(), name string) (bool, error) {
+	return s.add(schedule, cmd, name)
+}
+
+func (s *Scheduler) Start(ctx context.Context) {
+	s.logger.Printf("Starting CRON scheduling with [%d] jobs.", len(s.jobs))
```

Comment:
Out of curiosity, wouldn't using a `ticker` for each cron job simplify the code?
https://gobyexample.com/tickers
You could just convert the cron regex to a numeric value (amount of seconds) and then have the ticker sleep for that amount of time

Reply:
I do not think it will necessarily simplify the code more than what it is, the body of it gorountine when dispacting a job will practically the same with replacing the time.sleep with a listening to the timer channel.